platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheValueSourceIntegrationTest.groovy

```diff
@@ -701,4 +704,139 @@ class ConfigurationCacheValueSourceIntegrationTest extends AbstractConfiguration
+            gradle.allprojects {
+                tasks.register('check', ValueTask) {
+                    value = sharedValue
+                }
+            }
+        """
+
+        when:
+        configurationCacheRun 'check'
+
+        then:
+        output.count('The values are the same') == 2
+    }
+
+    def "value source can be shared across build services"() {
+        given:
+        createDirs 'foo', 'bar'
+        settingsFile """
+            import org.gradle.api.provider.*
+
+            rootProject.name = 'root'
+            include 'foo', 'bar'
+
+            abstract class RandomValueSource implements ValueSource<Integer, ValueSourceParameters.None> {
+                @Override Integer obtain() {
+                    new Random().nextInt()
+                }
+            }
+
+            abstract class ValueProviderService implements ${BuildService.name}<Parameters>{
+
+                interface Parameters extends ${BuildServiceParameters.name} {
+                    Property<Integer> getValue()
+                }
+
+                Integer getValue() {
+                    parameters.value.get()
+                }
+            }
+
+            abstract class ValueCheckerService implements ${BuildService.name}<${BuildServiceParameters.name}.None>{
+
+                private Object value = null
+
+                synchronized def check(Object o) {
+                    if (value === null) {
+                        value = o
+                    } else {
+                        // value should be boxed once and cached
+                        assert value === o
+                        println 'The values are the same'
+                    }
+                }
+            }
+
+            abstract class ValueTask extends DefaultTask {
+                @ServiceReference('valueProvider') abstract Property<ValueProviderService> getValueProvider()
+                @ServiceReference('valueChecker') abstract Property<ValueCheckerService> getValueChecker()
+                @TaskAction def check() {
+                    valueChecker.get().check(valueProvider.get().value)
+                }
+            }
+
+            def valueChecker = gradle.sharedServices.registerIfAbsent('valueChecker', ValueCheckerService) {}
+
+            def sharedValue = providers.of(RandomValueSource) {}
+            gradle.allprojects {
+                def provider = gradle.sharedServices.registerIfAbsent(name + 'ValueProvider', ValueProviderService) {
+                    parameters {
+                        value = sharedValue
+                    }
+                }
+                tasks.register('check', ValueTask) {
+                    valueProvider = provider
+                }
+            }
+        """
+
+        and:
+        def configurationCache = newConfigurationCacheFixture()
+
+        when:
+        configurationCacheRun 'check'
+
+        and:
+        configurationCacheRun 'check'
```

Comment:
Because we have other tests that validate value sources as inputs to tasks but I assumed there wasn't one for value sources as parameters to build services (that are then used by tasks).

Reply:
I made both tests run the build twice to assert on the value being obtained at execution time.